### DISCUSSION TITLE
Append and list access efficiency

### DISCUSSION THREAD

**Discussion by minyoung90** [sqlalchemy/sqlalchemy#7963] [Usage Questions] [answered]
I can't find any information about list append efficiency

if we have Parent, child models (1 to m relationship) then, we can do it like...

```python
# 1
parent.chldren.append(child)

# 2
parent.children[-1].name = "helloworld"
```

I am very curious about this code.
What I checked was that #2 query of debug message of sqlalchemy gave me two queries which were to get the first child index and the last child index. and #1 gave me nothing, but only update query. I couldn't find any query for memory caching about children.

So, my conclusion is... Even if I have 1 million children, the above append method will execute in constant time and list access doesn't have to do join for getting the certain position of the child (is it lazy load for only one child?).

Is it right?

**Comment by zzzeek** [comment, marked as answer]
the background on "memory caching" is a foundational concept of "relationship" which is that collections are persistently loaded until expired unless they are "dynamic".   background is at https://docs.sqlalchemy.org/en/14/orm/collections.html#working-with-large-collections  

the answer depends entirely on specifics:

1. is parent.children a dynamic relationship?
2. if the answer to `#1` is "no", is parent.children loaded in memory?   (Also we assume the little used "noload" strategy is not set)
3. if the answers are "no, no" then either line above, when first called, will emit SQL to load *all* rows into memory, so if that's 1M rows, that's 1M entries.  you wouldnt want that to happen.
4. if parent.children is a dynamic relationship, then the first operation will emit no SQL, the second will emit a query with a LIMIT on it

basically if the collection has 1M rows you'd need to use "dynamic" or "noload".

**Reply by minyoung90** [reply]
@zzzeek Thanks for great answer!

**Comment by minyoung90** [comment]
So, I checked carefully, there were "memory caching" because I didn't use "lazy=dynamic"
